jax/_src/extend/ffi.py

```diff
@@ -140,3 +148,110 @@ def _ir_attribute(obj: Any) -> ir.Attribute:
     elif isinstance(mlir_type, ir.FloatType):
       return ir.FloatAttr.get(mlir_type, obj)
   raise TypeError(f"Unsupported attribute type: {type(obj)}")
+
+
+def ffi_call(
+    target_name: str,
+    result_shape_dtypes: DuckTypedArray | Sequence[DuckTypedArray],
+    *args: ArrayLike,
+    vectorized: bool = False,
+    **kwargs: Any,
+) -> Array | Sequence[Array]:
+  """Call a foreign function interface (FFI) target.
+
+  Like :func:`~jax.pure_callback`, the behavior of ``ffi_call`` under
+  :func:`~jax.vmap` depends on the value of ``vectorized``. When ``vectorized``
+  is ``True``, the FFI target is assumed to satisfy: ``ffi_call(xs) ==
+  jnp.stack([ffi_call(x) for x in xs])``. In other words, calling the FFI target
+  with an extra leading dimension should return the same result as calling it
+  within a loop and stacking along the zeroth axis. Therefore, the FFI target
+  will be called directly on batched inputs (where the batch axes are the
+  leading dimensions). Additionally, the callbacks should return outputs that
+  have corresponding leading batch axes. If ``vectorized`` is ``False`` (the
+  default behavior), transforming this ``ffi_call`` under :func:`~jax.vmap` will
+  result in a :func:`~jax.lax.scan` with the ``ffi_call`` in the body.
+
+  Args:
+    target_name: the name of the XLA FFI custom call target that was registered
+      using :func:`~jaxlib.xla_client.register_custom_call_target`.
+    result_shape_dtypes: an object, or sequence of objects, with ``shape`` and
+      ``dtype`` attributes which are expected to match the shape and dtype of
+      the custom call output or outputs. :class:`~jax.ShapeDtypeStruct` is often
+      used to define the elements of ``result_shape_dtypes``.
+    *args: the arguments passed to the custom call.
+    vectorized: boolean specifying whether the callback function can operate in
+      a vectorized manner, as described above.
+    **kwargs: keyword arguments that are passed as named attributes to the
+      custom call using XLA's FFI interface.
+
+  Returns:
+    One or more :class:`~jax.Array` objects whose shapes and dtypes match
+    ``result_shape_dtypes``.
+  """
+  if isinstance(result_shape_dtypes, Sequence):
```

Comment:
Have you considered always returning multiple results: `result_shape_dtypes` should always be a sequence, and the restult is always a sequence. The typing declarations would be simpler and the usage also.

Reply:
Good question! I did consider that, but decided that it was probably worth supporting single outputs since a large fraction of the uses of `custom_call` that I could find in Google code, and across GitHub only produce a single output. Furthermore, `pure_callback` also supports both single and multiple outputs (Note: that's not really a fair comparison since it also supports arbitrary pytree output), so I aimed to support as much of that interface as we can for an FFI. So, since the implementation is fairly straightforward, I came down on the side of supporting single or multiple outputs transparently, but I don't feel strongly if there is pressure otherwise!

Reply:
SGTM